packages/playground/website/builder/index.js

```diff
@@ -0,0 +1,982 @@
+
+	if (
+		prevKey.length === 3 &&
+		prefix.length >= 3 &&
+		prevKey.join('<') === 'slug<pluginZipFile<steps'
+	) {
+		const wpParams = new URLSearchParams();
+		wpParams.set('action', 'query_plugins');
+		wpParams.set('request[page]', '1');
+		wpParams.set('request[per_page]', '200');
+		wpParams.set('request[locale]', 'en_US');
+		wpParams.set('request[search]', prefix);
+		wpParams.set('request[wp_version]', '6.4');
+		const proxyParams = new URLSearchParams();
+
+		proxyParams.set(
+			'url',
+			`http://api.wordpress.org/plugins/info/1.2/?${wpParams}`
+		);
+
+		if (debounce) {
+			clearTimeout(debounce);
+			debounce = null;
+		}
+
+		document.body.setAttribute('data-loading', true);
+
+		debounce = setTimeout(async () => {
+			try {
+				const res = await fetch(
+					`https://playground.wordpress.net/plugin-proxy.php?${proxyParams}`
+				);
+				const json = await res.json();
+				json?.plugins.forEach((p) => {
+					var doc = new DOMParser().parseFromString(
+						p.name,
+						'text/html'
+					);
+					const meta = doc.documentElement.textContent;
+					callback(null, [
+						{
+							name: p.slug,
+							value: qA + p.slug + qB,
+							score: 1,
+							meta,
+						},
+					]);
+				});
+			} finally {
+				document.body.setAttribute('data-loading', false);
+			}
+		}, 250);
+	}
+
+	if (
+		prevKey.length === 3 &&
+		prefix.length >= 3 &&
+		prevKey.join('<') === 'slug<themeZipFile<steps'
+	) {
+		const wpParams = new URLSearchParams();
+		wpParams.set('action', 'query_themes');
+		wpParams.set('request[page]', '1');
+		wpParams.set('request[per_page]', '200');
+		wpParams.set('request[locale]', 'en_US');
+		wpParams.set('request[search]', prefix);
+		wpParams.set('request[wp_version]', '6.4');
+		const proxyParams = new URLSearchParams();
+		proxyParams.set(
+			'url',
+			`http://api.wordpress.org/themes/info/1.2/?${wpParams}`
+		);
+
+		if (debounce) {
+			clearTimeout(debounce);
+			debounce = null;
+		}
+
+		document.body.setAttribute('data-loading', true);
+
+		debounce = setTimeout(async () => {
+			try {
+				const res = await fetch(
+					`https://playground.wordpress.net/plugin-proxy.php?${proxyParams}`
+				);
+				const json = await res.json();
+				json?.themes.forEach((p) => {
+					var doc = new DOMParser().parseFromString(
+						p.name,
+						'text/html'
+					);
+					const meta = doc.documentElement.textContent;
+					callback(null, [
+						{
+							name: p.slug,
+							value: qA + p.slug + qB,
+							score: 1,
+							meta,
+						},
+					]);
+				});
+			} finally {
+				document.body.setAttribute('data-loading', false);
+			}
+		}, 250);
+	}
+
+	switch (prevKey[0]) {
+		case 'preferredVersions': {
+			const used = await getPrevSiblings(editor, pos);
+			list.push(...['wp', 'php'].filter((s) => !used.includes(s)));
+			break;
+		}
+
+		case 'wp':
+			list.push('latest');
+			break;
+
+		case 'php':
+			list.push(...(await completePhpVersion(prefix)));
```

Comment:
Let's ship this as is, but for posterity:

This approach will break sooner than later. It introduces tight coupling between the schema and the editor. However, we'll keep iterating on this schema. We'll add, remove, and change Blueprint properties every month and I would really like the editor to just keep working without any maintenance and housekeeping chores involved.

Let's ship this PR to give folks the editor they're waiting for and create a follow-up issue to decouple the schema from the editor.

Reply:
@seanmorris good question! VS Code seems to be able to autocomplete based on an arbitrary schema. My thinking is this:

* We know "where" in the JSON we are, as in we know the top-level key (such as `steps`), the key below it (such as `[2]`, and so on.
* We know what the schema says about that path we're in. It says the top-level ref is `"#/definitions/Blueprint"`, 
* We can follow that and see there's an object definition under `definitions` > `Blueprint`, and then we can see it has data under `properties` > `steps` which is an array of `"#/definitions/StepDefinition"`.
* We can resolve that ref to see what properties are expected there and how they're discriminated based on `"propertyName": "step"`

The only bit of custom logic here would be requesting the wp.org API to autocomplete the plugins and themes names.

Reply:
@adamziel 
Doesn't this code do that?
https://github.com/WordPress/wordpress-playground/pull/773/files#diff-52c7ea8f333f6dfa85a4351f19e1ab5d1f1d56fcbe7e5ff3bd82803afbb8a994R10
https://github.com/WordPress/wordpress-playground/pull/773/files#diff-52c7ea8f333f6dfa85a4351f19e1ab5d1f1d56fcbe7e5ff3bd82803afbb8a994R30

Reply:
You're right, this PR ships generic functions that can read from specific places in that schema.

My point is that this PR also ships functions that are strongly tied to a specific structure – like that `switch` statement this thread is attached to where keys like `wp`, `php`, `steps`, and `features` are explicitly listed out and tied to schema-specific autocompletion functions like `completePhpVersion()` or `completeSteps()`.

If the JSON schema changes, this program will require changes, too. At the same time, the same logic could be expressed in a way that works for any JSON schema.